### ISSUE TITLE
Package.resolve is't updated after update Package.swift

### ISSUE THREAD

**Issue by ra1028** [renovatebot/renovate#6924]
**What Renovate type are you using?**

Renovate for GitHub

**Describe the bug**

Renovate supports Swift Package Manager, but `Package.resolved` isn't updated after `Package.swift` is updated.
That's a lock file of Swift tool dependencies, so it should be updated if it exists on git, otherwise, always diff will be made on the local environment after only `Package.swift` is updated. 

**Relevant debug logs**

There is no such thing as a log, as it's a specification failure, not a runtime error.

**To Reproduce**

This happens whenever there is an update to the Swift package listed in `Package.swift`.

**Comment by rarkins** [comment]
Thanks, this feature is not yet supported. Please assist by providing:
- Link(s) to where this is best documented
- Any guides to how this can be run under Linux
- A public repo that reproduces the problem/requirement

**Comment by ra1028** [comment]
@rarkins 
Hi, thanks for your response.
I've created a sample public repository about this issue.

> Link(s) to where this is best documented

renovate swift document: https://docs.renovatebot.com/modules/manager/swift
sample pr: https://github.com/ra1028/sample-renovate6924/pull/1  

> Any guides to how this can be run under Linux

Just run `swift package update` after `Package.swift` is updated on the image that Swift lang is installed.

> A public repo that reproduces the problem/requirement

- repo: https://github.com/ra1028/sample-renovate6924
- pr: https://github.com/ra1028/sample-renovate6924/pull/1

**Comment by rarkins** [comment]
Definitely update and not resolve?


Also I was hoping for Swift's own documentation rather than ours, plus any examples of running it in Ubuntu 18.04 under Docker, or something close.

**Comment by ra1028** [comment]
@rarkins 

As you can see, `Package.resolved` is not changed after `Package.swift` is updated.
https://github.com/ra1028/sample-renovate6924/pull/1

Swift package manager documentation: 
- https://github.com/apple/swift-package-manager/blob/master/Documentation/Usage.md#resolving-versions-packageresolved-file

**Comment by viceice** [comment]
I think we need to call `swift package resolve` this will only update the resolved file, while `swift package update` updates the swift file too with latest available updates

**Comment by rarkins** [comment]
I think it should be swift package resolve and not swift package update

**Comment by viceice** [comment]
Official docker images: https://hub.docker.com/_/swift

**Comment by viceice** [comment]
docker file:
https://github.com/apple/swift-docker/blob/bcaa665ab185e9af7d82612c9436950e92466451/5.2/ubuntu/18.04/Dockerfile

we need to update our buildpack and a new gitrepo for `docker-swift`

**Comment by viceice** [comment]
then renovate needs to extract tools version from comment to use correct swift version

```
// swift-tools-version:5.3
```

https://github.com/ra1028/sample-renovate6924/blob/a11e2e7eb7445db0d4df0e1982b78c1a5c4398ae/Package.swift#L1

**Comment by viceice** [comment]
We can use git-tags as datasource with a regex versioning [swift-5.2.5-RELEASE](https://github.com/apple/swift/releases/tag/swift-5.2.5-RELEASE)

I think we need use `stabillityDays=1` to wait for build binaries.

**Comment by rarkins** [comment]
New repo: https://github.com/renovatebot/docker-swift

Buildpack issue: https://github.com/renovatebot/docker-buildpack/issues/50

**Comment by Ali-Dervisoglu** [comment]
Yes

**Comment by viceice** [comment]
which swift version we should build? Where to start? v5, v5.3, v4 ...

**Comment by viceice** [comment]
![image](https://user-images.githubusercontent.com/1798109/96739119-bc136100-13bf-11eb-8c6c-a05a7602a539.png)
 renovate won't be able to resolve under linux, because the test repo above explicit needs macos

**Comment by viceice** [comment]
> which swift version we should build? Where to start? v5, v5.3, v4 ...

I propose v5.3, as this supports updating older resolved files and supports ubuntu 20.04 focal

**Comment by viceice** [comment]
https://github.com/renovatebot/docker-buildpack/pull/56

**Comment by viceice** [comment]
blocked
- https://github.com/containerbase/base/issues/16
- https://github.com/renovatebot/renovate/issues/13683


swift installer needs an update

**Comment by viceice** [comment]
This can now continued, as the refactorings are done.

**Comment by Fab1n** [comment]
Is it possible to continue on that issue? It would be great. What was the issue with the closed PR?

**Comment by rarkins** [comment]
@Fab1n sure, please go ahead

**Comment by viceice** [comment]
> Is it possible to continue on that issue? It would be great. What was the issue with the closed PR?

the author didn't respond to our requested changes for a long time.

**Comment by eduardboamba** [comment]
Seems this issue was left in limbo. It's over 3 years since it was reported.

**Comment by joseewu** [comment]
hi 👋 any updates that we will proceed this PR? we are also looking forward to it 😄

**Comment by setchy** [comment]
Earlier today my Lead iOS Engineer raised this as being a useful feature addition to our Renovate workflow.

Is there any pre-req items I should be aware of for this feature?  I'd be happy to try and contribute it, although my interactions with Swift Package Manager are exactly `undefined` atm 😓

**Comment by rarkins** [comment]
Swift seems supported already in Containerbase: https://github.com/search?q=repo%3Acontainerbase%2Fbase%20swift&type=code

Hopefully this reproduction repo is still valid: https://github.com/ra1028/sample-renovate6924

Hopefully running `swift resolve` after a package file update is enough for a start.

A simple/similar one I added recently is Bun: https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/bun/artifacts.ts

The main difference here is that you probably don't want to treat it as a Buffer/binary file although even that might work.

**Comment by doozMen** [comment]
> I think it should be swift package resolve and not swift package update

ok so there seams to be confusion about this. It should be `swift package update` from the start. You should not suggest to update the Package.swift. Package.swift should be edited manually. The Package.swift file defines the minimum version a library is compatible with, the Package.resolved defines the version that is used during build. 

> Note: As a reference you could read [package resolved docs](https://github.com/apple/swift-package-manager/blob/main/Documentation/Usage.md#resolving-versions-packageresolved-file)

I will try to clarify why updating Package.swift for a dependency is not the way to go. The goal of the bot should be to detect wether there are updates available for the package. If versioning is done correctly it could suggest updates that can be done automatically. This is what `swift package update` does. It checks the package.swift file for how dependencies are defined and updates the once that are needed.

So the update process is handled by `swift package update`. 

`swift package resolve` will resolve to the latest version that it can use based on what is defined in the package.swift file. But it will not go fetch the dependencies to be ready for build.

So in short if you would want to make the bot check wether ar not there are updates it should

1. mv the package.resolved (if present) to a temp copy
2. do swift package resolve
3. diff the 2 resolved files
4. come up with a suggestion of update

The updated version should then be committed to the resolved file. 

When to update the Package.swift then?

When the library needs features that are in the next version of the dependency, effectively making it incompatible with packages that cannot yet update to that version number.

**Comment by doozMen** [comment]
Swift runs just fine in linux and you have installation options at https://www.swift.org/install/#:~:text=on%20download%20page.-,Linux,-Docker 

As for github actions the latest ubuntu has swift build in so you do not need to do anything.

**Comment by malcommac** [comment]
is swift still unsupported by renovatebot yet?

**Comment by ddanielczyk** [comment]
Any update on this, please?

**Comment by pixlwave** [comment]
Seems this was fixed in https://github.com/renovatebot/renovate/pull/41534
